ledger/store/src/helpers/memory/internal/nested_map.rs

```diff
@@ -0,0 +1,1372 @@
+                // Return an iterator over each key.
+                keys.into_iter().map(move |k| (Cow::Owned(m), Cow::Owned(bincode::deserialize(&k).unwrap())))
+            })
+            .collect_vec()
+            .into_iter()
+    }
+
+    ///
+    /// Returns an iterator over each confirmed value.
+    ///
+    fn values_confirmed(&'a self) -> Self::Values {
+        self.map_inner.read().clone().into_values().map(Cow::Owned)
+    }
+}
+
+/// Inserts the given map-key-value pair.
+fn insert<
+    M: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    V: Clone + PartialEq + Eq + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+>(
+    map: &mut BTreeMap<Vec<u8>, BTreeSet<Vec<u8>>>,
+    map_inner: &mut BTreeMap<Vec<u8>, V>,
+    m: &M,
+    k: &K,
+    v: V,
+) {
+    // Note: The 'unwrap' is safe here, because the maps and keys are defined by us.
+
+    // Serialize 'm'.
+    let m = bincode::serialize(m).unwrap();
+    // Serialize 'k'.
+    let k = bincode::serialize(k).unwrap();
+    // Concatenate 'm' and 'k' with a 0-byte separator.
+    let mk = [m.clone(), vec![0u8], k.clone()].concat();
+
+    map.entry(m).or_insert_with(BTreeSet::new).insert(k);
+    map_inner.insert(mk, v);
+}
+
+/// Removes the given map-key pair.
+fn remove_map<
+    M: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    V: Clone + PartialEq + Eq + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+>(
+    map: &mut BTreeMap<Vec<u8>, BTreeSet<Vec<u8>>>,
+    map_inner: &mut BTreeMap<Vec<u8>, V>,
+    m: &M,
+) {
+    // Note: The 'unwrap' is safe here, because the maps and keys are defined by us.
+
+    // Serialize 'm'.
+    let m = bincode::serialize(m).unwrap();
+    // Remove 'm'.
+    let keys = map.remove(&m);
+    // Iteratively remove the key-value pairs.
+    if let Some(keys) = keys {
+        for k in keys {
+            // Concatenate 'm' and 'k' with a 0-byte separator.
+            let mk = [m.clone(), vec![0u8], k].concat();
+            map_inner.remove(&mk);
+        }
+    }
+}
+
+/// Removes the given map-key pair.
+fn remove_key<
+    M: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    V: Clone + PartialEq + Eq + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+>(
+    map: &mut BTreeMap<Vec<u8>, BTreeSet<Vec<u8>>>,
+    map_inner: &mut BTreeMap<Vec<u8>, V>,
+    m: &M,
+    k: &K,
+) {
+    // Note: The 'unwrap' is safe here, because the maps and keys are defined by us.
+
+    // Serialize 'm'.
+    let m = bincode::serialize(m).unwrap();
+    // Serialize 'k'.
+    let k = bincode::serialize(k).unwrap();
+
+    map.entry(m.clone()).or_insert_with(BTreeSet::new).remove(&k);
```

Comment:
```suggestion
    map.entry(m.clone()).or_default().remove(&k);
```

Reply:
also, the `clone` can be avoided if `mk` was created as described before and before this call